jaxlib/mosaic/gpu/custom_call.cc

```diff
@@ -528,32 +577,109 @@ absl::StatusOr<std::tuple<void*, MosaicHostFunc*>> CachedCompileAndInit(
   return cache->at(key).GetHostLaunch();
 }
 
-void MosaicGPUCustomCall(void* stream, void** buffers, char* opaque,
-                         size_t opaque_len, XlaCustomCallStatus* status) {
-  if (reinterpret_cast<uintptr_t>(opaque) % alignof(KernelHash)) {
+std::pair<absl::flat_hash_set<int64_t> *, absl::Mutex *>
+GetNeedsNvshmemBarrier() {
+  static absl::Mutex mutex;
+  static auto &needs_nvshmem_barrier = *new absl::flat_hash_set<int64_t>;
+  return std::make_pair(&needs_nvshmem_barrier, &mutex);
+}
+
+absl::Status MosaicGpuExecute(gpuStream_t stream, ffi::RemainingArgs inputs,
+                              ffi::RemainingRets results,
+                              absl::string_view kernel_hash,
+                              absl::string_view module,
+                              bool use_custom_barrier, xla::RunId run_id) {
+  if (reinterpret_cast<const uintptr_t>(kernel_hash.data()) %
+          alignof(KernelHash) ||
+      kernel_hash.size() != sizeof(KernelHash)) {
     fprintf(stderr, "Misaligned opaque pointer\n");
     abort();
   }
-  auto hash = *reinterpret_cast<KernelHash*>(opaque);
+  auto hash = *reinterpret_cast<const KernelHash *>(kernel_hash.data());
   CUcontext ctx;
   if (cuCtxGetCurrent(&ctx) != CUDA_SUCCESS) {
     fprintf(stderr, "Failed to get current CUDA context\n");
     abort();
   }
   CacheKey key(hash, reinterpret_cast<uintptr_t>(ctx));
-  auto ctx_and_kernel = CachedCompileAndInit(key, opaque + sizeof(KernelHash));
-  if (!ctx_and_kernel.ok()) {
-    XlaCustomCallStatusSetFailure(status,
-                                  ctx_and_kernel.status().message().data(),
-                                  ctx_and_kernel.status().message().size());
-    return;
+  auto ctx_kernel_comm = CachedCompileAndInit(key, module.data());
+  if (!ctx_kernel_comm.ok()) {
+    return absl::InternalError(ctx_kernel_comm.status().message());
```

Comment:
Nit: just return the status as is without re-wrapping it in `absl::IternalError`?

I think we should have access to the `TF_ASSIGN_OR_RETURN` macro, that's also an option.